x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/Driver.java

```diff
@@ -449,17 +476,28 @@ public DriverProfile profile() {
      * @param extraIterations how many iterations to add to the previous status
      * @param status the status of the overall driver request
      */
-    private void updateStatus(long extraCpuNanos, int extraIterations, DriverStatus.Status status) {
+    private void updateStatus(long extraCpuNanos, int extraIterations, DriverStatus.Status status, String reason) {
         this.status.getAndUpdate(prev -> {
+            long now = System.currentTimeMillis();
+            DriverSleeps sleeps = prev.sleeps();
+            switch (status) {
+                case ASYNC, WAITING -> sleeps = sleeps.sleep(reason, now);
+                case RUNNING -> {
+                    if (prev.status() == DriverStatus.Status.ASYNC || prev.status() == DriverStatus.Status.WAITING) {
+                        sleeps = sleeps.wake(now);
+                    }
+                }
+            }
             return new DriverStatus(
                 sessionId,
                 startTime,
-                System.currentTimeMillis(),
+                now,
                 prev.cpuNanos() + extraCpuNanos,
                 prev.iterations() + extraIterations,
                 status,
                 statusOfCompletedOperators,
-                activeOperators.stream().map(op -> new DriverStatus.OperatorStatus(op.toString(), op.status())).toList()
+                activeOperators.stream().map(op -> new DriverStatus.OperatorStatus(op.toString(), op.status())).toList(),
+                sleeps
```

Comment:
My hope is that this isn't too expensive to do for every time we go async. I'm not worried about the case where we run for five minutes and update the status or run for a zillion iterations. This is tiny compared to that no matter how you slice it. But if we have an operator that goes async all the time then recording the sleep is non-trivial.